test/rateLimiters/slidingWindowCounter.test.ts

```diff
@@ -0,0 +1,476 @@
+            expect((await getWindowFromClient(client, user2)).currentTokens).toBe(1);
+            expect((await getWindowFromClient(client, user3)).currentTokens).toBe(user3Tokens);
+        });
+
+        test("sliding window doesn't allow capacity/window size <= 0", () => {
+            expect(() => new SlidingWindowCounter(0, 10, client)).toThrow(
+                'SlidingWindowCounter windowSize and capacity must be positive'
+            );
+            expect(() => new SlidingWindowCounter(-1, 10, client)).toThrow(
+                'SlidingWindowCounter windowSize and capacity must be positive'
+            );
+            expect(() => new SlidingWindowCounter(10, -1, client)).toThrow(
+                'SlidingWindowCounter windowSize and capacity must be positive'
+            );
+            expect(() => new SlidingWindowCounter(10, 0, client)).toThrow(
+                'SlidingWindowCounter windowSize and capacity must be positive'
+            );
+        });
+
+        test('all windows should be able to be reset', async () => {
+            const tokens = 5;
+            await setTokenCountInClient(client, user1, tokens, null, timestamp);
+            await setTokenCountInClient(client, user2, tokens, null, timestamp);
+            await setTokenCountInClient(client, user3, tokens, null, timestamp);
+
+            limiter.reset();
+
+            expect((await limiter.processRequest(user1, timestamp, CAPACITY)).success).toBe(true);
+            expect((await limiter.processRequest(user2, timestamp, CAPACITY - 1)).success).toBe(
+                true
+            );
+            expect((await limiter.processRequest(user3, timestamp, CAPACITY + 1)).success).toBe(
+                false
+            );
+        });
+    });
+
+    describe('SlidingWindowCounter correctly updates Redis cache', () => {
+        afterEach(() => {
+            client.flushall();
+        });
+
+        test('timestamp correctly updated in redis', async () => {
+            let redisData: RedisWindow;
+
+            // blocked request
+            await limiter.processRequest(user1, timestamp, CAPACITY + 1);
+            redisData = await getWindowFromClient(client, user1);
+            expect(redisData.fixedWindowStart).toBe(timestamp);
+
+            timestamp += 1000;
+            // allowed request
+            await limiter.processRequest(user2, timestamp, CAPACITY);
+            redisData = await getWindowFromClient(client, user2);
+            expect(redisData.fixedWindowStart).toBe(timestamp);
+        });
+
+        test('current/previous tokens correctly updated in redis', async () => {
+            let redisData: RedisWindow;
+
+            await limiter.processRequest(user1, timestamp, 2);
+
+            redisData = await getWindowFromClient(client, user1);
+
+            expect(redisData.currentTokens).toBe(2);
+
+            await limiter.processRequest(user1, timestamp + WINDOW_SIZE + 1, 3);
```

Comment:
Can we also test `timestamp + WINDOW_SIZE` since that is the start of the next window? (e.g. 10 millisecond window size and previous start is 0 then end of the window is at 10 seconds not 11). Left a comment in the implementation about this.